### DISCUSSION TITLE
ARC-0043: Extending the Puzzle to a Full SNARK

### DISCUSSION THREAD

**Discussion by howardwu** [ProvableHQ/ARCs#77] [ARCs]
# ARC-0043: Extending the Puzzle to a Full SNARK

---
arc: 0043
title: Extending the Puzzle to a Full SNARK
authors: @howardwu 
discussion: #77 
topic: Protocol
status: Draft
created: October 15, 2024
---

## Abstract

This ARC proposes extending the current blockchain puzzle from a synthesis-focused operation to a full Succinct Non-interactive Argument of Knowledge (SNARK). The primary goals are to incentivize the acceleration of a full SNARK on dedicated hardware and improve block verification times through succinct puzzle verification. This proposal aims to mitigate bottlenecks in the block verification pipeline caused by synthesis operations, and to leverage recent advancements in GPU and FPGA acceleration to improve the puzzle challenge for provers.

## Background

SNARK generation has seen significant performance improvements over the past several years, in part due to efforts such as the Aleo Testnets and ZPrize. While proof generation was initially bottlenecked by algorithms such as Multi-Scalar Multiplication (MSM) and Number Theoretic Transform (NTT), these algorithms have been accelerated significantly, resulting in synthesis becoming the most time consuming operation. With recent improvements to synthesis resulting from the launch of the Aleo Mainnet, there is now an opportunity to further optimize the end-to-end proving and verifying process.

## Proposal

This proposal extends the current puzzle to a full SNARK with the following key changes:

1. Increase the number of instructions in the puzzle
2. Expand the types of instructions and operands
3. Implement a zkSNARK for verification

### 1. Increased Instruction Count

The current puzzle has an artificially constrained number of instructions due to limitations in consensus verification. By implementing a full SNARK, one can expand the instruction count by orders of magnitude. This will significantly harden the puzzle for provers, expanding the utility of dedicated hardware and improving on current computational requirements for proving.

### 2. Expanded Instruction Set

Introducing a zkSNARK for verification allows for the inclusion of richer opcodes with larger operands. This expansion of the instruction set will increase the complexity and versatility of the puzzle challenge. However, careful consideration must be given to instruction selection to avoid degeneracy of operands (e.g., operands becoming 0 and propagating through the remainder of the function).

### 3. zkSNARK Verification

Implementing a zkSNARK for verification will enable succinct puzzle proof verification. This will improve block verification times, enhancing overall network performance.

## Technical Considerations

1. The extension will be built on the existing proof system and elliptic curve to minimize technical risk and debt in snarkOS and snarkVM.
2. The existing synthesizer is capable of expanding both the number and types of instructions in the Varuna SNARK.
3. To maintain non-malleability and non-amortizability, the puzzle will be extended to a SNARK, not a zkSNARK. This is crucial to prevent grindability on a single proof, which would simplify the problem for provers.

## Benefits

1. Incentivizes the development and use of dedicated hardware for full SNARK acceleration.
2. Significantly improves block verification times through succinct puzzle verification.
3. Enhances puzzle complexity and security by allowing for a much larger instruction set.
4. Leverages recent advancements in GPU and FPGA acceleration for synthesis operations.

## Risks and Mitigations

**Risk:** Increased complexity in puzzle design and verification.
**Mitigation:** Thorough testing and gradual implementation, with potential for a phased rollout.

**Risk:** Potential for new attack vectors due to expanded instruction set.
**Mitigation:** Rigorous security analysis & audits, alongside careful selection of new instructions.

**Risk:** Increased hardware requirements for miners/provers.
**Mitigation:** Phased implementation to allow for gradual hardware upgrades.

## Conclusion

Extending the current puzzle to a full SNARK incentivizes a significant advancement for proving hardware. It addresses current bottlenecks for consensus, improves overall network performance, and incentivizes hardware acceleration for cryptographic operations. The next step is to gather community feedback to refine and improve this proposal.

**Comment by apruden2008** [comment]
Thanks. I am supportive in principle of expanding the scope of the puzzle in order to make it more “useful”. 

What are the implications of this change on the economics? In other words, are there any changes expected in the puzzle difficulty **adjustment** algorithm that would potentially affect the inflation rate of ALEO?

**Comment by mingo6666666666** [comment]
May I ask when the algorithm upgrade is expected to **start**?

**Reply by howardwu** [reply]
This ARC is currently a proposal in draft stage. As discussions and design proposals mature, this ARC will converge on a concrete design that can be implemented, analyzed, and reviewed.

**Comment by zkxuerb** [comment]
I think this is the right direction both for incentivising the improvement of dedicated hardware for cryptographic operations and very importantly improving network performance. 

Do we have an idea on what the current impact of puzzle verification is to network consensus as a bottleneck ( in terms of time to verify ) ?

**Reply by vicsn** [reply]
> Do we have an idea on what the current impact of puzzle verification is to network consensus as a bottleneck ( in terms of time to verify ) ?

Because of the high verification requirements (0.5 GiB memory, takes 100s of ms), `MAX_SOLUTIONS` per block is set to only 4 at the moment to ensure solutions don't slow down consensus. If the puzzle was replaced by a full snark, this limitation could be removed, as verification of a `Transmission::Solution` would be as fast to verify as an (execution) `Transmission::Transaction`.

**Comment by Ryan-MetaworksLab** [comment]
Hi, we are Metaworks Lab from Australia. We are keen to contribute to Aleo and hope to work with the community to participate in the improvement and implementation of the [ARC0043](https://github.com/AleoNet/ARCs/discussions/77) proposal. 

We work with the ZK hardware accelerator team to do software and hardware co-design. We have the ability to support ARC43 and can also do general ZK accelerator. We can do:
1. Increase the number of instructions in the puzzle, expand the types of instructions and operands
      - We can implement operation instructions and data types that are compatible with the CPU and GPU versions, and can run efficiently in parallel
2. Generate zkSNARK proofs for puzzles
     - After executing the R1CS of the puzzle, use the generated Constraints to generate the prove of KZG Polynomial Commitments, where MSM is accelerated by CUDA
    - Later, FPGA and ASIC dedicated to Aleo puzzle solving can be provided
3. zkSNARK verification of puzzles
    - Verify the verify of KZG Polynomial Commitments, MSM is accelerated on the CPU using the Pippenger algorithm

> We are Metaworks Lab from Australia, with 8 full-time ZKP R&D personnel in Shanghai. We have been actively participating in the testnet and mainnet mining of the Aleo project, with strong computing power support and optimized mining algorithms.

**Comment by LoiusZK** [comment]
Hi, as a loyal fan of Aleo, I have a few questions to ask:

1. How much will the TPS be improved after the implementation of this PR? Do you have internal test data?
2. If the block time is stabilized at about 10 seconds, can the daily transactions volume reach tens of millions?
3. Will the block capacity or transaction capacity affect TPS?
4. How long will it take to implement this PR? Is there a time plan and arrangement?

The current TPS is indeed very bad and needs to be solved urgently, otherwise the ecosystem cannot develop. Thanks.

**Reply by howardwu** [reply]
@LoiusZK The most straightforward way to increasing TPS is by introducing more validators into the network. This ARC is likely to have little impact on TPS, as consensus currently verifies at most 4 valid puzzle solutions per block, each at around 100-200ms. This ARC would likely reduce the amount of time needed, but could also enable more puzzle solutions to be submitted in each block, allowing for more provers to earn rewards.

**Reply by LoiusZK** [reply]
@howardwu If simply increase the number of validators based on the current instruction set puzzle algorithm, won’t the block time also increase? Cause we know the final calculation result of the current instruction set puzzle algorithm is not a Proof, so the validators need to execute it individually to verify, and cannot execute Proof verification in batches, which will also affect TPS. I think this ARC should also be a preparation for adding more validators in the future right? 
According to what you said, the maximum number of solutions will increase to more than 4 for per block, which is of great concern to the provers, but I am more concerned about the original intention of proposing this ARC, that is,  and the original intention of optimizing and updating the puzzle algorithm in the future. My understanding is that the current and future optimization of the puzzle algorithm is to accelerate zkvm and improve TPS. In fact, for Aleo, the biggest difficulty is to take into account the output of PoW+PoS in line with the expected planning of the economic model, while ensuring that there is a high enough TPS to support the development of a large-scale ecosystem.

**Reply by StarLI-Trapdoor** [reply]
> @LoiusZK The most straightforward way to increasing TPS is by introducing more validators into the network. This ARC is likely to have little impact on TPS, as consensus currently verifies at most 4 valid puzzle solutions per block, each at around 100-200ms. This ARC would likely reduce the amount of time needed, but could also enable more puzzle solutions to be submitted in each block, allowing for more provers to earn rewards.

It's CPU performance to verify one valid puzzle with 100~200ms? I think if GPU is used, the verification period can be much less.

**Comment by d0cd** [comment]
While all three improvements to the puzzle are desirable, it appears that **Goal #3: implementing a SNARK for verification** is the most important. Without it, increasing the instruction count and complexity would only increase the solution verification time. In [these](https://github.com/AleoNet/snarkVM/blob/staging/ledger/puzzle/epoch/docs/spec.md#benchmarks) benchmarks, we find that a puzzle program with 1000 instructions results in a solution that takes ~500ms to verify. Ideally, we'd like to increase instruction count well beyond this number. With succinct solution verification, we can increase program complexity in meaningful (and virtually unbounded) ways.

An added benefit is that the protocol can support a larger number of solutions per block, allowing rewards to be distributed to more provers.

One open question is: **how do validators synthesize the verifying key?** Furthermore, how do clients, which may have lower compute resources get this key? Solutions are verified live during consensus. Furthermore, the puzzle program changes every epoch. Even for small programs, the overall circuit can contain millions of constraints. Synthesizing verifying keys in these settings can take a significant amount of time. A naive design is to require validators to synthesize the key at the beginning of the epoch, however this could result in a verification delay.

A possible solution is introduce a buffer during which a validator can synthesize the key. For example, the puzzle program can be generated using the previous epoch hash, while the seed for the puzzle inputs can use the current epoch hash. Validators are then responsible for storing and distributing verification keys to clients. 

Curious to hear what people think and other approaches to consider!

**Reply by howardwu** [reply]
Thanks for the thoughts @d0cd. Do you see any possibility of having a pipelined design, whereby provers have 1 full epoch to synthesize the verifying key and get it to all validators, then 1 full epoch to find solutions for the puzzle? By pipelining these two stages, one can have a continuous puzzle and verifying key synthesis procedure.

**Reply by d0cd** [reply]
Pipelining sounds like a promising approach. I imagine validators would still have to verify the submitted verifying key against `Certificate`. How much cheaper is it to verify a certificate versus create a verifying key, given an `Assignment`? Tangentially, do you envision that prover key submission is an incentivized component of the puzzle design?

**Reply by vicsn** [reply]
> How much cheaper is it to verify a certificate versus create a verifying key, given an Assignment?

Verification of just the Verifying Key `Certificate` is cheap, roughly constant (~10 ms assuming few public variables), so that should not be an issue.

> Tangentially, do you envision that prover key submission is an incentivized component of the puzzle design?

Very cool idea! I'm in favour to explore. To be clear, I think it would only make sense for **verifying key** submission to be an incentivised component, as this is what validators need. Currently I don't see any issues with this, though note that 1. it increases complexity, so can be added in a later stage 2. it adds a liveness assumption, namely that we expect a network participant to submit a verifying key to a validator participating in consensus.

**Comment by c00567056** [comment]
When will the algorithm upgrade be completed? Can you give a plan?

**Reply by howardwu** [reply]
This ARC is currently a proposal in draft stage. As discussions and design proposals mature, this ARC will converge on a concrete design that can be implemented, analyzed, and reviewed.

**Reply by c00567056** [reply]
thank you very much!

**Reply by MarioMoreschini1992** [reply]
Hello, came by this ARC in the provers section of Aleo Developer, then we as provers have just 1 main question: will this ARC impact the current ASIC that we have (mainly goldshell and iceriver) algorithm? Will they still working or the change proposed can affect the ability of this server to propose solutions to the network?

**Comment by Hugo-cflabs** [comment]
@howardwu

We at Computation Frontier Labs are a team researching computation acceleration technology, including zero-knowledge proofs.

We believe this proposal demonstrates a positive trend for Aleo in constructing a useful ZKP prover platform. It's beneficial for promoting the development and adoption of ZKP technology, much like Aleo's efforts with ZPrize.

Furthermore, we have some questions:

1. Considering machine development and life cycles, especially for ASIC machines, what's the rough timeline for this proposal—1-2 quarters or 1-2 years?
2. As we've estimated, will Aleo's new puzzle prover support Varuna's zkSNARK to accelerate client-side proving? Is there a roadmap for this?
3. Is it possible for Aleo to further upgrade the AVM to support other ISAs and more universal ZKVM technology, making the prover useful for other ZK applications?

PS: We've published an article analyzing the Aleo project: https://medium.com/@CFrontier_Labs/the-journey-toward-aleos-universal-zkvm-1df9114fbd86

We hope Aleo continues to encourage the community to develop ZKVM-specific hardware machines. This could benefit not only Aleo but also other ZKP projects, as the same acceleration technology and machines could be applied across various ZKP projects.

**Comment by LoiusZK** [comment]
@howardwu 


> Increase the number of instructions in the puzzle
> Expand the types of instructions and operands

Regarding this ARC, I have a few questions to ask or my suggestions are as follows:

1. Is the increase in the type and number of instructions within the range of 60 instructions of AVM? Or will new instructions and instruction types be added?
2. Is the generation of new instruction sequences random or fixed?

> By pipelining these two stages, one can have a continuous puzzle and verifying key synthesis procedure.


3. I agree with this pipeline because as the algorithm becomes more and more complex, it is necessary to set aside some time to process it. I suggest generating a program or procedure at least one epoch in advance.
4. How much will the instruction count of each epoch program increase? Is it 10 times? Or 100 times?
5. What is the basis for determining the number of instructions? Once the number of instructions is determined, does the number of instructions in each epoch change? Or is it fixed?

**Comment by Csmcg55889900** [comment]
> # ARC-0043：将谜题扩展为完整的SNARK
> ## 弧：0043
> 首页外文书人文历史Extensioning the Puzzle to a Full SNARK
> 作者： @Howardwu
> 讨论： #77
> 主题：协议
> 状态：草稿
> 创建时间：2024年10月15日
> ## 摘要
> 这个ARC提出将当前的区块链难题从以合成为中心的操作扩展到完整的简洁非交互式知识论证（Simplnct Non-interactive Argument of Knowledge,SNARK）。主要目标是激励在专用硬件上加速完整的SNARK，并通过简洁的谜题验证来改善块验证时间。该提案旨在缓解由合成操作引起的块验证流水线中的瓶颈，并利用GPU和FPGA加速的最新进展来改善证明者的难题挑战。
> 
> ## 背景信息
> SNARK一代在过去几年中取得了显著的性能改进，部分原因是Aleo Testnets和ZPrize等的努力。虽然证明生成最初被多标量乘法（MSM）和数论变换（NTT）等算法瓶颈，但这些算法已经显著加速，导致合成成为最耗时的操作。随着Aleo Mainnet的推出，最近对合成进行了改进，现在有机会进一步优化端到端证明和验证过程。
> 
> ## 建议书
> 这个提议将当前的谜题扩展到一个完整的SNARK，主要变化如下：
> 
> 1. 增加拼图中的指令数量
> 2. 扩展指令和操作数的类型
> 3. 实现zkSNARK进行验证
> 
> ### 1.增加指令计数
> 由于一致性验证的限制，当前的难题具有人为限制的指令数量。通过实现完整的SNARK，可以将指令计数扩展到数量级。这将大大强化证明者的难题，扩大专用硬件的效用，并改进当前证明的计算需求。
> 
> ### 2.扩展指令集
> 引入zkSNARK进行验证允许包含具有较大操作数的更丰富操作码。指令集的这种扩展将增加谜题挑战的复杂性和多功能性。但是，必须仔细考虑指令的选择，以避免操作数的简并（例如，操作数变为0并通过函数的其余部分传播）。
> 
> ### 3.zkSNARK验证
> 实现zkSNARK进行验证将实现简洁的难题证明验证。这将改善块验证时间，提高整体网络性能。
> 
> ## 技术考虑
> 1. 该扩展将建立在现有的证明系统和椭圆曲线的基础上，以最大限度地减少snarkOS和snarkVM的技术风险和债务。
> 2. 现有的合成器能够扩展Varuna SNARK中的指令数量和类型。
> 3. 为了保持非可出售性和非可摊销性，谜题将扩展到SNARK，而不是zkSNARK。这对于防止单一证明上的可磨性至关重要，这将简化证明者的问题。
> 
> ## 优势
> 1. 激励开发和使用专用硬件进行全SNARK加速。
> 2. 通过简洁的谜题验证，显著改善了块验证时间。
> 3. 通过允许更大的指令集来增强谜题的复杂性和安全性。
> 4. 利用GPU和FPGA加速方面的最新进展来进行合成操作。
> 
> ## 风险和缓解
> **风险：** 增加谜题设计和验证的复杂性。 **缓解措施：** 彻底的测试和逐步实施，有可能分阶段推出。
> 
> **风险：** 由于扩展的指令集，新的攻击媒介的潜力。 **缓解措施：** 严格的安全分析和审计，以及仔细选择新的指令。
> 
> **风险：** 矿工/开发人员的硬件需求增加。 **缓解措施：** 分阶段实施，允许逐步升级硬件。
> 
> ## 结论
> 将当前的谜题扩展到完整的SNARK可以激励硬件证明的重大进步。它解决了当前共识瓶颈，提高了整体网络性能，并激励了加密操作的硬件加速。下一步是收集社区反馈，以完善和改进此提案。

很好

**Comment by Wen-wen999** [comment]
您好，作为Aleo的忠实拥护者，我相信隐私之王Aleo未来的应用，一定可以改变世界web3时代，有问题想请教您，如果采用ARC—0043，我们基于当前主网算法开发ASIC芯片机，是否需要等一等，还是采用FPGA矿机推动Aleo发展？

**Comment by QUINCYchoy** [comment]
do it mean current mining machines will become obsolete?

**Reply by vicsn** [reply]
All current communications by Howard on online fora and presentations implied that the old puzzle and old machines will remain used if ARC-0043 becomes a reality
